### PR TITLE
fix(FEC-10844): advanced caption settings custom caption is always marked even when choose sample

### PR DESCRIPTION
**Pull request by Yuvalke**
### Description of the Changes

Issue: the condition of || doesn't work properly with 0 as the current value and always takes the default option.
Solution: create the default values as a static getter to be able to use it on static methods and on the first initialize of the properties.
merge the setting and default value to make sure all set correctly.
add relevant tests.

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/track/text-style.js [hunk 1]
```diff
@@ -138,17 +140,31 @@ class TextStyle {
...
+    let textStyle = new TextStyle();
+    textStyle.fontEdge = currentSetting.fontEdge;
+    textStyle.fontSize = currentSetting.fontSize;
+    textStyle.fontScale = currentSetting.fontScale;
+    textStyle.fontColor = currentSetting.fontColor;
+    textStyle.fontOpacity = currentSetting.fontOpacity;
+    textStyle.backgroundColor = currentSetting.backgroundColor;
+    textStyle.backgroundOpacity = currentSetting.backgroundOpacity;
+    textStyle.fontFamily = currentSetting.fontFamily;
+    return textStyle;
```
Comment (OrenMe) [inline src/track/text-style.js:167]: In general I think the better approach here is to use a helper method to check if settings exist and not null or undefined:
```javascript
function get(value, default) {
  let result = default;
  if (value && value !== null && value !== undefined) {
    result = value;
  }
  return result;
}


let textStyle = new TextStyle();
textStyle.fontEdge = get(setting.fontEdge, clonedTextStyle.fontEdge);
textStyle.fontSize = get(setting.fontSize, clonedTextStyle.fontSize);
textStyle.fontScale = get(setting.fontScale, clonedTextStyle.fontScale);
textStyle.fontColor = get(setting.fontColor, clonedTextStyle.fontColor);
textStyle.fontOpacity = get(setting.fontOpacity, clonedTextStyle.fontOpacity);
textStyle.backgroundColor = get(setting.backgroundColor, clonedTextStyle.backgroundColor);
textStyle.backgroundOpacity = get(setting.backgroundOpacity, clonedTextStyle.backgroundOpacity);
textStyle.fontFamily = get(setting.fontFamily, clonedTextStyle.fontFamily;
return textStyle;

```

and if we take it further, there's a new JS feature called [nullish coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) that does exactly this:

```javascript
textStyle.fontEdge = setting.fontEdge ?? clonedTextStyle.fontEdge;
```

and there's a babel plugin for it: @babel/plugin-proposal-nullish-coalescing-operator

**Review by OrenMe — CHANGES_REQUESTED** [review]

File: src/track/text-style.js [hunk 2]
```diff
@@ -138,17 +140,31 @@ class TextStyle {
     return 'rgba(' + color.concat(opacity).join(',') + ')';
   }
 
+  static get defaultValues(): PKTextStyleObject {
```
[the remark this answers is not in the document]
Comment (Yuvalke) [reply src/track/text-style.js:143]: why not? if the user wants to know what is the default value?

[the remark this answers is not in the document]
Comment (OrenMe) [reply src/track/text-style.js:143]: There are too many options to get default value. You can also create a new textStyle and get value from it.
If you have a justified reason for new api feel free to raise it.

File: src/track/text-style.js (hunk 1 above)
Comment (Yuvalke) [reply src/track/text-style.js]: this one looks better to you?? I thought it will be better to merge cause it will take the value if exists

Comment (OrenMe) [reply src/track/text-style.js]: If all of the ECMA script experts that write the JavaScript spec think this is good and this is a supported feature already in browsers(check chrome dev tools now if you want 😀 ) then it is good enough to me.

Comment (Yuvalke) [reply src/track/text-style.js]: I mean the get helper :)

File: src/track/text-style.js [hunk 3]
```diff
@@ -139,16 +139,23 @@ class TextStyle {
...
-    clonedTextStyle.backgroundColor = setting.backgroundColor || clonedTextStyle.backgroundColor;
-    clonedTextStyle.backgroundOpacity = setting.backgroundOpacity || clonedTextStyle.backgroundOpacity;
-    clonedTextStyle.fontFamily = setting.fontFamily || clonedTextStyle.fontFamily;
-    return clonedTextStyle;
+    const getValue = (newValue: any, defaultValue: any): any => {
+      let value = defaultValue;
+      if (typeof newValue !== 'undefined' && newValue !== null) {
+        value = newValue;
+      }
+      return value;
```
Comment (RoyBregman) [inline src/track/text-style.js:147]: can be replace with a one liner:
 `return (typeof newValue !== 'undefined' && newValue !== null) ? newValue : defaultValue`
 
 But its a matter of taste...

File: src/track/text-style.js [hunk 4]
```diff
@@ -139,16 +139,23 @@ class TextStyle {
...
+    };
+    let textStyle = new TextStyle();
+    textStyle.fontEdge = getValue(setting.fontEdge, textStyle.fontEdge);
+    textStyle.fontSize = getValue(setting.fontSize, textStyle.fontSize);
+    textStyle.fontScale = getValue(setting.fontScale, textStyle.fontScale);
+    textStyle.fontColor = getValue(setting.fontColor, textStyle.fontColor);
+    textStyle.fontOpacity = getValue(setting.fontOpacity, textStyle.fontOpacity);
+    textStyle.backgroundColor = getValue(setting.backgroundColor, textStyle.backgroundColor);
+    textStyle.backgroundOpacity = getValue(setting.backgroundOpacity, textStyle.backgroundOpacity);
+    textStyle.fontFamily = getValue(setting.fontFamily, textStyle.fontFamily);
```
Comment (RoyBregman) [inline src/track/text-style.js:157]: Can this maybe replaced with an iteration?

File: test/src/player.spec.js [hunk 5]
```diff
@@ -1216,14 +1216,46 @@ describe('Player', function () {
...
-        textStyle.backgroundColor = TextStyle.StandardColors.RED;
-        textStyle.fontColor = TextStyle.StandardColors.CYAN;
         textStyle.fontEdge = TextStyle.EdgeStyles.RAISED;
+        textStyle.fontSize = '75%';
+        textStyle.fontScale = '3';
+        textStyle.fontColor = TextStyle.StandardColors.BLACK;
+        textStyle.fontOpacity = TextStyle.StandardOpacities.SEMI_HIGH;
+        textStyle.backgroundOpacity = TextStyle.StandardOpacities.SEMI_LOW;
+        textStyle.fontFamily = TextStyle.FontFamily.ARIAL;
+        textStyle.backgroundColor = TextStyle.StandardColors.RED;
```
Comment (RoyBregman) [inline test/src/player.spec.js:1226]: one of these tests is with 0 value?
the black color?

Comment (Yuvalke) [reply test/src/player.spec.js:1226]: Transperncy

Comment (Yuvalke) [reply test/src/player.spec.js:1226]: I used all the options to make sure I didn't miss something

File: src/track/text-style.js (hunk 4 above)
Comment (Yuvalke) [reply src/track/text-style.js:157]: if we'll get a new non-relevant key/private key it'll add to this set.

File: src/track/text-style.js (hunk 3 above)
Comment (RoyBregman) [reply src/track/text-style.js:147]: So unless you wanna change this to my suggestion then it is approved

**Review by RoyBregman — APPROVED** [review]

**Pull request merged**